backend/cmd/server.go

```diff
@@ -86,6 +97,102 @@ func main() {
 	})
 }
 
+// GetContextKeyAndContext returns Kcontext , ContextKey for using these in CacheMiddleWare function.
+// It also return span and ctx that will help while using handleError function.
+func GetContextKeyAndKContext(w http.ResponseWriter,
+	r *http.Request, c *HeadlampConfig) (context.Context,
+	trace.Span, string, *kubeconfig.Context, error,
+) {
+	ctx := r.Context()
+	ctx, span := telemetry.CreateSpan(ctx, r, "cluster-api", "handleClusterAPI",
+		attribute.String("cluster", mux.Vars(r)["clusterName"]),
+	)
+
+	defer span.End()
+
+	contextKey, err := c.getContextKeyForRequest(r)
+	if err != nil {
+		c.handleError(w, ctx, span, err, "failed to get context Key:", http.StatusBadRequest)
+		return nil, nil, "", nil, err
+	}
+
+	kContext, err := c.KubeConfigStore.GetContext(contextKey)
+	if err != nil {
+		c.handleError(w, ctx, span, err, "failed to get context", http.StatusNotFound)
+		return nil, nil, "", nil, err
+	}
+
+	return ctx, span, contextKey, kContext, nil
+}
+
+// CacheMiddleWare is Middleware for Caching purpose. It involves generating key for a request,
+// authorizing user , store resource data in cache and returns data if key is present.
+func CacheMiddleWare(c *HeadlampConfig) mux.MiddlewareFunc { //nolint:funlen
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx, span, contextKey, kContext, err := GetContextKeyAndKContext(w, r, c)
+			if err != nil {
+				c.handleError(w, ctx, span, err, "failed to get context and Kcontext", http.StatusNotFound)
+				return
+			}
+
+			if kContext.Error != "" {
+				c.handleError(w, ctx, span, errors.New(kContext.Error), "context has error", http.StatusBadRequest)
+				return
+			}
+
+			if r.Method != "GET" {
+				// For POST, PUT, DELETE, directly proxy the request and purge local data.
+				next.ServeHTTP(w, r)
+				key, _ := k8cache.GenerateKey(r.URL, contextKey)
+
+				k8cache.DeleteKeys(key, k8scache)
+
+				return
+			}
+
+			rcw := k8cache.CreateResponseCapture(w)
+
+			key, err := k8cache.GenerateKey(r.URL, contextKey)
+			if err != nil {
+				c.handleError(w, ctx, span, err, "failed to generate key ", http.StatusBadRequest)
+				return
+			}
+
+			isAllowed, authErr := k8cache.IsAllowed(r.URL, kContext, w, r)
+			if authErr != nil {
+				k8cache.StoreAfterAuthError(k8scache, isAllowed, next, key, w, r, rcw)
+
+				return
+			} else if !isAllowed && k8cache.CheckForURL(r.URL.Path) {
+				_ = k8cache.ReturnAuthErrorResponse(w, r, contextKey)
+
+				return
+			}
+
+			served, err := k8cache.LoadFromCache(k8scache, isAllowed, key, w, r)
+			if err != nil {
+				c.handleError(w, ctx, span, errors.New(kContext.Error), "failed to load from cache", http.StatusServiceUnavailable)
+			}
+
+			if served {
+				c.telemetryHandler.RecordEvent(span, "Served from cache")
+				return
+			}
+
+			go k8cache.CheckForChanges(k8scache, contextKey, r, rcw)
```

Comment:
The CheckForChanges shouldn't de initiated from the cache middleware. If 10 users request 10 resources, we will have 100 go routines checking for changes. This should be called outside the cache middleware.